components/push/src/communications.rs

```diff
@@ -140,6 +144,38 @@ impl ConnectHttp {
         };
         Ok(headers)
     }
+
+    fn handle_response_error(&self, response: &viaduct::Response) -> error::Result<()> {
+        // An error response, the extended object structure is retrieved from
+        // https://autopush.readthedocs.io/en/latest/http.html#response
+        #[derive(Deserialize)]
+        struct ResponseError {
+            pub errno: u32,
+            pub message: String,
+        }
+        if response.is_server_error() {
+            let response_error = response.json::<ResponseError>()?;
+            return Err(CommunicationServerError(format!(
+                "General Server Error: {}",
+                response_error.message
+            ))
+            .into());
+        }
+        if response.is_client_error() {
+            let response_error = response.json::<ResponseError>()?;
+            if response.status == status_codes::CONFLICT {
+                return Err(AlreadyRegisteredError.into());
+            }
+            if response.status == status_codes::GONE && response_error.errno == UAID_NOT_FOUND_ERRNO
+            {
+                return Err(UAIDNotRecognizedError(response_error.message).into());
+            }
+            return Err(
+                CommunicationError(format!("Unhandled client error {:?}", response)).into(),
+            );
+        }
+        Ok(())
```

Comment:
For a moment here, I thought you were silently ignoring other kinds of errors by returning `Ok(())`, but IIUC you will *always* call this `handle_response_error` method and the fall-through case here is the case that no error occurred. That seems fine, but it may be worth considering a name of the method that makes this behaviour more clear.

Reply:
Good point! I renamed it to `check_response_error` 😄